### PR TITLE
Scaffold Generator integration into CDK

### PR DESCRIPTION
**Pull request by JonasSchaub**
Integrates the Scaffold Generator functionalities (described [here](https://doi.org/10.26434/chemrxiv-2022-7tf0h)) into the new CDK "scaffold" package. 
Scaffold Generator offers the generation of different molecular scaffolds, scaffold trees, and scaffold networks.

### CODE REVIEW

**Comment by egonw** [comment]
@JonasSchaub, I think this PR has not been on my radar recently :(

**Comment by JonasSchaub** [comment]
@egonw That's fine, no worries ;)

**Comment by JonasSchaub** [comment]
Hey @egonw, please wait with reviewing the code a little longer. We just got reviews for the accompanying manuscript we submitted and we might have to change something in the code. I will come back to you when these things have been sorted.

**Comment by JonasSchaub** [comment]
Hi @egonw , 
I have now submitted the revised manuscript (if you are interested: https://doi.org/10.26434/chemrxiv-2022-7tf0h-v2) and changed the code accordingly. 
I very much hope that the manuscript gets accepted in this form and I won't have to change anything else (in the code). 
It's maybe best if I tell you once the manuscript has been accepted and you start reviewing the code here only then.
Best regards,
Jonas

**Comment by JonasSchaub** [comment]
Hi @egonw ,
the manuscript has been accepted in its current state!
So, no additional code changes will be necessary at this point. You can review the pull request now whenever you find time.
Thanks and kind regards,
Jonas

File: tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/NetworkNode.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+ * The parents are the smaller parent scaffolds.
+ * @param <MoleculeType> As MoleculeType, any data type can be defined.
+ *                     In our scenario, the node contains a CDK IAtomContainer.
+ *
+ * @author Julian Zander, Jonas Schaub (zanderjulian@gmx.de, jonas.schaub@uni-jena.de)
+ * @version 1.0.0.0
+ */
+public class NetworkNode <MoleculeType> extends ScaffoldNodeBase<MoleculeType> {
+
+    //<editor-fold desc="Private variables">
```
Comment (egonw) [inline tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/NetworkNode.java]: Please remove editor suggestion lines

File: tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/NetworkNode.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    //</editor-fold>
+
+    //<editor-fold desc="Public methods">
+    @Override
+    public boolean isOrphan() {
+        return parents.isEmpty();
+    }
+
+    /**
+     * @throws NullPointerException if parameter is null
```
Comment (egonw) [inline tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/NetworkNode.java:70]: Missing description. All public methods should have full JavaDoc.

Comment (egonw) [reply tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/NetworkNode.java:70]: Applies to other places too. Please check all code.

File: tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/NetworkNode.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) 2022 Julian Zander <zanderjulian@gmx.de>
```
Comment (egonw) [inline tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/NetworkNode.java:2]: Thanks!

File: tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java [hunk 4]
```diff
@@ -0,0 +1,2412 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.TreeMap;
+
+/**
+ * This class is designed to generate different molecule scaffolds and frameworks.
+ * It contains several other methods for the decomposition of molecules.
+ * All details of the functionality and implementation are described <a href="https://doi.org/10.26434/chemrxiv-2022-7tf0h-v2">
```
Comment (egonw) [inline tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java]: Please update.

File: tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java [hunk 5]
```diff
@@ -0,0 +1,2412 @@
...
+ * here</a>.<p>
+ *
+ * Furthermore, the molecules can be decomposed according to the <a href="https://doi.org/10.1021/ci600338x">
```
Comment (egonw) [inline tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java:64]: There is `@cdk.cite` for citing literature, but these needs revamping anyway. @johnmay, this is fine?

File: tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java [hunk 6]
```diff
@@ -0,0 +1,2412 @@
...
+ * The Scaffold Generator library is described in <a href="https://doi.org/10.1186/s13321-022-00656-x"> Schaub et al.
+ * "Scaffold Generator: a Java library implementing molecular scaffold functionalities in the Chemistry Development Kit
+ * (CDK)" (J Cheminform 14, 79, 2022)</a>
+ *
+ * @author Julian Zander, Jonas Schaub (zanderjulian@gmx.de, jonas.schaub@uni-jena.de)
+ * @version 1.0.4.0
+ */
+public class ScaffoldGenerator {
+
+    //<editor-fold desc="Enum">
```
Comment (egonw) [inline tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java]: Plz check all code. See above.

File: tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java [hunk 7]
```diff
@@ -0,0 +1,2412 @@
...
+    public static final Aromaticity AROMATICITY_MODEL_SETTING_DEFAULT = new Aromaticity(ElectronDonation.cdk(), Cycles.cdkAromaticSet());
+
+    /**
+     * Default setting for which SmilesGenerator should be used.
+     * By default, unique SMILES are used.
+     */
+    public static final SmilesGenerator SMILES_GENERATOR_SETTING_DEFAULT = new SmilesGenerator(SmiFlavor.Unique | SmiFlavor.UseAromaticSymbols);
+
+    /**
+     * Default setting for whether rule 7 should be applied. By default, rule 7 is applied.
```
Comment (egonw) [inline tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java]: Rule 7 of what?

File: tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java [hunk 8]
```diff
@@ -0,0 +1,2412 @@
...
+        }
+        return tmpScaffoldNetwork;
+    }
+
+    /**
+     * Generates a network for each molecule in the list and merges the networks together. <p>
+     *
+     * Iteratively removes the terminal rings of each molecule. All resulting scaffolds of a molecule are saved in a ScaffoldNetwork.
+     * A new level is created when the total number of rings decreases by 1.
+     * Level 0 is the one with the fewest (mostly 1) rings and therefore the root. <br>
```
Comment (egonw) [inline tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java]: What do you mean with "mostly 1" here?

**Review by egonw — CHANGES_REQUESTED** [review]
I had a first look at the source code. I did not try to run the code yet.

I added a few comments.

File: tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java (hunk 5 above)
Comment (johnmay) [reply tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java:64]: Fine with me here

**Comment by johnmay** [comment]
This should probably be on the cdk repo rather than your fork Egon right? My one comment would be similar to the MAYGEN this code should not be depending on SMILES generator. Fine if it's a tool/app but for library functionality that should be decoupled.

File: tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java (hunk 5 above)
Comment (egonw) [reply tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java:64]: Actually, thinking about https://github.com/cdk/cdk/issues/931 I just commented on, if I manage to factor out the JavaDoc taglets then we can do more. We can make the source code more fair, and make the citations available as machine readable code.

**Comment by egonw** [comment]
> This should probably be on the cdk repo rather than your fork Egon right? 

Eventually, yes. I promised to look at it first. (But it's already in pretty good shape.)

> My one comment would be similar to the MAYGEN this code should not be depending
> on SMILES generator. Fine if it's a tool/app but for library functionality that should be decoupled.

It would be okay if the library part is in the new `structgen` module, but the application side
in the `tools/xxx` module?

**Comment by johnmay** [comment]
tools/scaffold then maybe a new module in app(s)/ I can’t remember what I called it but high level/convenient depict is done a bit a like this. 

Anyways all looks fine I just think it’s useful to make the coupling as loose as possible. When a lot of new dependencies are added to a module/pom it’s normally a sign it’s doing too much. 

Unless it’s designed in from the start it can be tricky to tease out the API from an end-to-end tool

File: tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/NetworkNode.java (hunk 3 above)
Comment (JonasSchaub) [reply tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/NetworkNode.java:2]: For putting the copyright notice? Welcome!

File: tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/NetworkNode.java (hunk 1 above)
Comment (JonasSchaub) [reply tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/NetworkNode.java]: Editor folds were removed with the latest commit.

File: tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java (hunk 6 above)
Comment (JonasSchaub) [reply tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java]: Editor folds were removed everywhere with the latest commit.

File: tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/NetworkNode.java (hunk 2 above)
Comment (JonasSchaub) [reply tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/NetworkNode.java:70]: With the latest commit, all public methods (also the overridden ones) should have a full JavaDoc.

File: tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java (hunk 4 above)
Comment (JonasSchaub) [reply tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java]: Has been updated on the latest commit.

File: tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java (hunk 5 above)
Comment (JonasSchaub) [reply tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java:64]: Sorry, I'm a bit lost here. Should I use the @cdk.cite annotation or not (yet)? And if yes, how exactly do I do it?

File: tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java (hunk 7 above)
Comment (JonasSchaub) [reply tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java]: I have added more details in the latest commit.

File: tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java (hunk 8 above)
Comment (JonasSchaub) [reply tool/scaffold/src/main/java/org/openscience/cdk/tools/scaffold/ScaffoldGenerator.java]: I have added more details on the latest commit.

**Comment by JonasSchaub** [comment]
Dear @egonw and @johnmay ,
first of all thank you for having a look at this!

> My one comment would be similar to the MAYGEN this code should not be depending on SMILES generator. Fine if it's a tool/app but for library functionality that should be decoupled.

If possible, I would very much like to keep the SMILES generator. It would be complicated to remove it and it has been explicitly described in the accompanying publication. 
Should I move everything to a different package, then? I am not exactly sure what your conclusion is on this matter.

**Comment by johnmay** [comment]
It's probably fine just note the different between a tool vs library.

@egonw I think one blocker is this looks like it's using Junit4 and we are now on 5 so that will need to be updated.

**Comment by JonasSchaub** [comment]
> It's probably fine just note the different between a tool vs library.

Thanks, I will keep this in mind. 

> I think one blocker is this looks like it's using Junit4 and we are now on 5 so that will need to be updated.

The pom of the scaffold module does not specify a JUnit version and it is therefore taken from the general CDK pom, is it not? 
This fork/branch is a bit older. The JUnit version will be updated "automatically" with the merge, right? 
Should I maybe test the code with JUnit 5 beforehand already?

**Comment by johnmay** [comment]
I think if you rebase on main it will break, the junit5 switch was not that long ago. It’s not too hard to migrate, IntelliJ makes this very easy.

**Comment by JonasSchaub** [comment]
I have now test-wise updated to JUnit 5 locally and did some adjustments to the test code. The main thing was simply replacing  usages of the Assert class with the Assertions class. 
The only other thing is that I cannot use "@Category(SlowTest.class)" anymore. Is it ok if I replace it like this with code in the test method itself?

````
CDKTestCase tmpTestCase = new CDKTestCase();
Assumptions.assumeTrue(tmpTestCase.runSlowTests());
````

I found this in https://github.com/cdk/cdk/blob/main/base/test-standard/src/test/java/org/openscience/cdk/ringsearch/AllRingsFinderTest.java but I cannot extend CDKTestCase with the test class because I am already extending the ScaffoldGenerator class. 

It seems like there are no additional adjustments needed.

**Comment by johnmay** [comment]
Thanks - How "slow" is the slow test?

https://github.com/cdk/cdk/blob/40b552eb70d2eafce9540b424f22bd0f710693d9/legacy/src/test/java/org/openscience/cdk/qsar/descriptors/atomic/IPAtomicLearningDescriptorTest.java#L109-L111

```java
@Test
@Tag("SlowTest")
void testIPDescriptor1() throws Exception {
...
```

**Comment by JonasSchaub** [comment]
The methods I tagged as slow tests are still quite fast, the "slowest" one takes about 0.5 seconds in IntelliJ on my notebook. Is it even necessary to tag them like this, then? Egon suggested it because they each import and parse 7 MOL files from resources.

Thanks for the code example!

**Comment by JonasSchaub** [comment]
Hi @egonw and @johnmay,
coming back to this after some time, do I still need to do some adjustments here or do you think that this branch is ready to be merged?

**Comment by egonw** [comment]
Moved to the CDK repo. Going to do some final checks there. Deep apologies for the delay.

**Pull request closed without merging**